google-ads-migration-examples/src/main/java/com.google.ads.googleads.migration/createcampaign/CreateCompleteCampaignAdWordsApiOnly.java

```diff
@@ -0,0 +1,429 @@
+    throws RemoteException, UnsupportedEncodingException {
+    // Get the AdGroupCriterionService.
+    AdGroupCriterionServiceInterface adGroupCriterionService =
+      adWordsServices.get(session, AdGroupCriterionServiceInterface.class);
+
+    AdGroupCriterionOperation[] operations = new AdGroupCriterionOperation[keywordsToAdd.size()];
+
+    for (int i = 0; i < keywordsToAdd.size(); i++) {
+      // Create the keyword.
+      Keyword keyword = new Keyword();
+      keyword.setText(keywordsToAdd.get(i));
+      keyword.setMatchType(KeywordMatchType.EXACT);
+
+      // Create the biddable ad group criterion.
+      BiddableAdGroupCriterion keywordBiddableAdGroupCriterion = new BiddableAdGroupCriterion();
+      keywordBiddableAdGroupCriterion.setAdGroupId(adGroup.getId());
+      keywordBiddableAdGroupCriterion.setCriterion(keyword);
+
+      // You can optionally provide these field(s).
+      keywordBiddableAdGroupCriterion.setUserStatus(UserStatus.PAUSED);
+
+      String encodedFinalUrl = String.format("http://example.com/mars/cruise/?kw=%s",
+        URLEncoder.encode(keyword.getText(), UTF_8.name()));
+      keywordBiddableAdGroupCriterion.setFinalUrls(new UrlList(new String[]{encodedFinalUrl}));
+
+      // Create the operation.
+      AdGroupCriterionOperation keywordAdGroupCriterionOperation = new AdGroupCriterionOperation();
+      keywordAdGroupCriterionOperation.setOperand(keywordBiddableAdGroupCriterion);
+      keywordAdGroupCriterionOperation.setOperator(Operator.ADD);
+
+      operations[i] = keywordAdGroupCriterionOperation;
+    }
+
+
+    // Add the keywords.
+    AdGroupCriterionReturnValue result = adGroupCriterionService.mutate(operations);
+
+    // Display the results.
+    for (AdGroupCriterion adGroupCriterionResult : result.getValue()) {
+      System.out.printf("Keyword ad group criterion with ad group ID %d, criterion ID %d, "
+          + "text '%s', and match type '%s' was added.%n",
+        adGroupCriterionResult.getAdGroupId(),
+        adGroupCriterionResult.getCriterion().getId(),
+        ((Keyword) adGroupCriterionResult.getCriterion()).getText(),
+        ((Keyword) adGroupCriterionResult.getCriterion()).getMatchType());
+    }
+    return result.getValue();
+  }
+
+  /**
+   * Creates text ads.
+   * @param adWordsServices the Google AdWords services interface.
+   * @param session the client session.
+   * @param adGroup the ad group for the text ad.
+   * @throws RemoteException if the API request failed due to other errors.
+   */
+  private AdGroupAd[] createTextAds(AdWordsServicesInterface adWordsServices,
+                                    AdWordsSession session, AdGroup adGroup)
+    throws RemoteException {
+
+    // Get the AdGroupAdService.
+    AdGroupAdServiceInterface adGroupAdService =
+      adWordsServices.get(session, AdGroupAdServiceInterface.class);
+
+    List<AdGroupAdOperation> operations = new ArrayList<>(NUMBER_OF_ADS);
```

Comment:
The style guide states that we should not specify the size of a list here. I can dig up the research if you're interested in the details.

Reply:
Fixed.